### PR TITLE
Fix resumed training drawing the first epoch with sampler epoch 0 when the dataloader uses workers

### LINKED ISSUE #21938: First epoch after resuming from checkpoint uses sampler epoch 0 (regression from #20775, num_workers > 0)
### Bug description

Since `2.5.3`, the first training epoch after resuming from an end-of-epoch checkpoint (`trainer.fit(..., ckpt_path=...)`) iterates the training dataloader with the sampler's epoch 0 permutation instead of the restored epoch, whenever the `DataLoader` uses worker processes (`num_workers > 0`). Later epochs of the resumed run are correct.

With `use_distributed_sampler=True` the Trainer controls `DistributedSampler.set_epoch`, so users cannot work around this from `LightningModule` code; every chained/preemptible DDP job replays the epoch-0 shuffle on its first epoch. Combined with `limit_train_batches` this becomes a data-coverage bug; each restart re-trains on the same fixed slice of the dataset.

### Cause

`FitLoop.setup_data()` creates the training iterator before any sampler epoch is set:

```python
# src/lightning/pytorch/loops/fit_loop.py, setup_data()
self._data_fetcher = _select_data_fetcher(trainer, RunningStage.TRAINING)
self._data_fetcher.setup(combined_loader)
iter(self._data_fetcher)  # creates the iterator inside the fetcher
```

`_set_sampler_epoch(dl, self.epoch_progress.current.processed)` only runs later, in `FitLoop.on_advance_start()`. A multi-process `DataLoader` iterator pulls its first indices from the sampler at creation, so the permutation is already drawn with `sampler.epoch == 0` (a fresh process) by the time `set_epoch` is called.

Before #20775 this did not matter for map-style datasets, because `_TrainingEpochLoop.on_run_start` re-created the iterator (after `set_epoch`) at the start of every epoch of a resumed run:

```python
# 2.5.2
if self.trainer.current_epoch > 0 and not self.restarting:
    iter(data_fetcher)
```

#20775 (2.5.3, "Fix double iteration bug when resumed from a checkpoint") added `and not self.trainer.fit_loop.is_resuming` to stop `IterableDataset`s from skipping data. As a side effect, the epoch-0 iterator from `setup_data()` is now the one consumed by the first resumed epoch:

```python
# 2.5.3 … master
if self.trainer.current_epoch > 0 and not self.trainer.fit_loop.is_resuming and not self.restarting:
    iter(data_fetcher)
```

`FitLoop.restarting` is coerced to `False` for end-of-epoch checkpoints, so `is_resuming` is the deciding term.

### Proposed fix

Set the sampler epoch before the initial iterator is created, so the fix in #20775 keeps its intent:

```python
# fit_loop.py, setup_data()
for dl in combined_loader.flattened:
    _set_sampler_epoch(dl, self.epoch_progress.current.processed)
self._data_fetcher = _select_data_fetcher(trainer, RunningStage.TRAINING)
self._data_fetcher.setup(combined_loader)
iter(self._data_fetcher)
```

(`epoch_progress.current.processed` is already the epoch about to train at that point; it is the value `on_advance_start` uses.)

### What version are you seeing the problem on?

v2.6

### Reproduced in studio

https://lightning.ai/sam-lab/templates/first-epoch-after-resuming-from-checkpoint-uses-sampler-epoch-0-regression-from-20775-num-workers-0~01m20cggj23c2dpnztwxz4ab7n

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.6.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

cc @ethanwharris

### PR DESCRIPTION
**Pull request by VihaanAgarwal**
## What does this PR do?

Since #20775 the first epoch of a run resumed from an end-of-epoch checkpoint iterates the training dataloader with the sampler's epoch 0 permutation whenever the `DataLoader` uses worker processes. `FitLoop.setup_data()` creates the iterator before `on_advance_start()` sets the sampler epoch, and a multi-process iterator prefetches its first indices at creation. #20775 stopped the epoch loop from recreating that iterator on resume, so the epoch-0 permutation is the one that gets trained on.

The sampler epoch is now set in `setup_data()` before the iterator is created, using the same `epoch_progress.current.processed` that `on_advance_start()` uses. Later epochs are unchanged.

`test_evaluation_loop_sampler_set_epoch_called` counted one `set_epoch` call per epoch; it now sees the extra call made at dataloader setup, so its expectation is updated.

Fixes #21938

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? Yes, #21938
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? Not needed
- Did you write any **new necessary tests**? Yes, `test_resume_sets_sampler_epoch_before_workers_draw_indices` fails on master with `[0] == [1]` and passes with the fix
- [x] Did you verify new and **existing tests pass** locally with your changes? `tests/tests_pytorch/loops` passes (246 passed, 3 skipped, macOS CPU)
- Did you list all the **breaking changes** introduced by this pull request? None
- Did you **update the CHANGELOG**? Yes

</details>

### CODE REVIEW

**Comment by mukeshr06** [comment]
Thanks for working on #21938. Would you welcome an additional regression test using DistributedSampler to compare the actual batch order in the first resumed epoch against the expected restored-epoch permutation, including persistent_workers=True and limit_train_batches? I would be happy to contribute that test coverage alongside this fix.

**Comment by VihaanAgarwal** [comment]
Thanks for the offer. The test here already pins the thing that broke: the epoch the sampler reads inside `__iter__`, which is where `DistributedSampler` picks it up to seed the permutation. A batch-order assertion would check the same fact one level down.

I'd rather keep this PR to the one behavior change so it stays easy to review. If you want to add the DistributedSampler case with `persistent_workers=True`, a follow-up once this lands would be a good place for it.
